File: python/Q53_maximum_subarray.py
#
# @lc app=leetcode id=53 lang=python3
#
# [53] Maximum Subarray
#

# @lc code=start
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        current: int = 0
        result: int = nums[0]

        for n in nums:
            if current < 0:
                current = 0

            current += n
            if current > result:
                result = current

        return result


# @lc code=end

File: rust/src/leetcode/Q53_maximum_subarray.rs
/*
 * @lc app=leetcode id=53 lang=rust
 *
 * [53] Maximum Subarray
 */

// @lc code=start
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut result: i32 = i32::MIN;
        let mut current: i32 = i32::MIN;

        for num in nums {
            if current < 0 {
                current = num;
            } else {
                current += num;
            }

            result = result.max(current);
        }

        result
    }
}
// @lc code=end
